pkg/apis/build/v1alpha1/source.go

```diff
@@ -8,10 +8,30 @@ import (
 	corev1 "k8s.io/api/core/v1"
 )
 
+// PruneOption defines the supported options for image pruning
+type PruneOption string
+
+const (
+	// Do not delete image after it was pulled
+	PruneNever PruneOption = "Never"
+
+	// Delete image after it was successfully pulled
+	PruneAfterPull PruneOption = "AfterPull"
+)
+
 // BundleContainer describes the source code bundle container to pull
 type BundleContainer struct {
 	// Image reference, i.e. quay.io/org/image:tag
 	Image string `json:"image"`
+
+	// Prune specifies whether the image is suppose to be deleted. Allowed
+	// values are 'Never' (no deletion) and `AfterPull` (removal after the
+	// image was successfully pulled from the registry).
+	//
+	// If not defined, it defaults to 'Never'.
+	//
+	// +optional
+	Prune *PruneOption `json:"prune,omitempty"`
```

Comment:
The change in a vaccum LGTM.

Disclaimer - my upcoming question stems from a lack of time to review the bundle related EPs and scenarios.  So please be gentle / apologies for any distraction.  Though sometimes these type of questions from non-SMEs sheds light on how users my interpret :-)  But in case it doesn't, I'll still post a prow "approve" in the interim for RH API signoff.

Does it make sense to control pruning of BundleContainer images at "higher" (or maybe just different) levels of our API?

I'm wondering what happens when we get more robust around volumes, caching, etc. and maybe we don't want to prune an image that is stored at a location shared by multiple BuildRun's until those BuildRuns are pruned, and hence this could be a future add on to the automatic buildrun cleanup that is also in flight with #1027 and we want API specification there.

Certainly all of that ^^ could be the provenance of a future EP, API change (however disruptive it might be) that we are fine with because we are still alpha level API.  And maybe even if ^^ makes sense it is just additive API and we still want what is here.

But anyway, just trying to probe and see what are y'all thoughts here.  Maybe this has already been considered and I just missed it.

thanks

Reply:
Interesting point @gabemontero. And is maybe considered already. Not 100 % sure as I am not seeing a direct relationship with volumes and caches.

So, basically we made `prune` here intentionally not a `bool` property, but an enumeration. The idea is that we can add more possible values in the future. The following two are what I personally had in mind:

1. `onSuccessfulBuildRunCompletion`. Here we delete the bundle image at the end of the BuildRun after all strategy steps were successful. It would basically end-up at the end of out image-post-processing step (that comes with shipwright managed push). This would address the scenario that one re-triggers a BuildRun which previously failed because of, for example, a flaky network and a rerun of the same source would be needed.
2. `onBuildRunDeletion`. Here we would tie the bundle lifetime with the BuildRun using some finalizer.

But extensibility is there at least in some way.

Reply:
We had something like this discussion already. I agree, what is proposed here is so to speak the minimum with regards to the prune implementation. The other interesting options would be what Sascha mentioned before, the deletion once the BuildRun completed successful, and of course deletion once the Build is deleted.

For the later two options, I considered that we would refactor the delete logic into a separate package.

I am up to suggestions here, we can open this PR to include the other two options already. We have it open now and have the context. Just let me know what you think.

Reply:
@HeavyWombat given we need this capability here in downstream, I personally would be in favor to not further extend it now. As long as we agree that we have an extensible approach, we can add other options and their implementation in the future. And the implementation of the other two options is not trivial today. The `onSuccessfulBuildRunCompletion` will become trivial once we have the shipwright managed push logic adding a step after the build strategy steps anyway. The `onBuildRunDeletion` option with the needed finalizer is a larger change to our logic and needs discussions/decisions on how we trigger the deletion in a finalizer with the necessary secret in the customer's namespace.

Reply:
Agreed.

Reply:
Thanks @gabemontero. I see what you mean. One could discuss that the setting would be inside `spec.retention`. There are probably pros and cons for both. Just some immediate thoughts:

- The lifecycle of the BuildRun and the source container image is decoupled because the source container image will always be created before the BuildRun. The `onBuildRunDeletion` will eventually couple the deletion part, but never the creation, at least from a pure Shipwright Build controller perspective. From an e2e user perspective, the user - like with the streaming approach - submits a BuildRun and the bundling and upload of the source container image and the creation of the BuildRun will mostly be one single user action. Client applications (CLIs) can even go further and completely hide the existence of the image from the user (that's what the Code Engine CLI does).
- The fields in `retention` do not mention any entity, nor does the `prune` field mention it. It is implicit by their parent field (and its documention) what they are applicable for (BuildRun, source container image). Moving the prune field into `retention` would require field renames. Further clarity would be to rename the Build's `spec.retention` to `spec.buildRunRetention`.
- Grouping all retention-related fields together helps those who come from / look for this perspective. Users who look at how to define the source would maybe miss a field that is related to the source container image but elsewhere in the spec.

Yeah, so in short, I think the field would not be mis-placed in retention, but I also think it fits into the source. I have a favor for the current approach, mainly because I like to have all fields for one source together (and I am aware that a very similar argument can be used for the other option). :-)

Reply:
yep let's leave it like you all have it here @SaschaSchwarze0 

we have a path for "connecting the pruning flow" via `onBuildRunDeletion` etc.

and let's wait to see *IF* we ever get feedback of the retention/pruning centric API ilk and then adjust

if I had to predict, if we get feedback at all along these lines, it might split along developer vs. operator/administrator line, but that is just a guess